File: 数据结构2/堆排序.cpp
#include <iostream>
#include <algorithm>

using namespace std;

const int N = 100010;

int n, m;
int h[N], size1;

void down(int u)
{
    int t = u;
    if (u * 2 <= size1 && h[u * 2] < h[t])
        t = u * 2;
    if (u * 2 + 1 <= size1 && h[u * 2 + 1] < h[t])
        t = u * 2 + 1;
    if (u != t)
        swap(h[u], h[t]);
    down(t);
}
void up(int u)
{
    while (u / 2 && h[u / 2] > h[u])
    {
        swap(h[u / 2], h[u]);
        u /= 2;
    }
}

int main()
{
    cin >> n;
    for (int i = 1; i <= n; i++)
        cin >> h[i];
    size1 = n;
    for (int i = n / 2; i != 0; i--)
        down(i);

    while (m--)
    {
        cout << h[1];
        h[1] = h[size1];
        size1--;
        down(1);
    }

    return 0;
}

File: 数据结构2/堆排序.js
const readline = require("readline");

rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
let n = 0,
  m = 0;
let h = [];
let size = 0;
let result = [];
//-----------------------------------------------------------------------------

rl.on("line", (data) => {
  if (n === 0) {
    [n, m] = data.split(" ").map((i) => +i);
  } else {
    main(data);
  }
});
//-----------------------------------------------------------------------------
function main(data) {
  h = [0, ...data.split(" ").map((i) => +i)];
  size = n;
  for (let i = Math.floor(n / 2); i; i--) down(i);
  while (m--) {
    result.push(h[1]);
    h[1] = h[size];
    size--;
    down(1);
  }
  console.log(result.join(" "));
}
function down(x) {
  let t = x;
  if (x * 2 <= size && h[x * 2] < h[t]) t = x * 2;
  if (x * 2 + 1 <= size && h[x * 2 + 1] < h[t]) t = x * 2 + 1;
  if (t != x) {
    swap(t, x);
    down(t);
  }
}
function swap(a, b) {
  let t = h[a];
  h[a] = h[b];
  h[b] = t;
}
